### PR TITLE
Update OpenSfM

### PR DESCRIPTION
**Pull request by pierotofy**
This PR brings in the latest changes to OpenSfM, which among other things include the `reprojection of derivatives which are now computed analytically instead of autodiff-ed (20%-45% speed-up on opensfm reconstruct)`

WIP

### CODE REVIEW

**Comment by smathermather** [comment]
w00t!

File: opendm/config.py [hunk 1]
```diff
@@ -743,7 +743,7 @@ def config(argv=None, parser=None):
                         type=float,
                         action=StoreValue,
                         metavar='<positive float>',
-                        default=15,
+                        default=10,
```
Comment (smathermather) [inline opendm/config.py:736]: 👍

**Comment by lucyllewy** [comment]
Would it be possible to get this PR included? :-) https://github.com/mapillary/OpenSfM/pull/658

**Comment by pierotofy** [comment]
Yep!

**Comment by lucyllewy** [comment]
also needs mapillary/OpenSfM#663 (I made a booboo mistake in the first PR)

**Comment by smathermather** [comment]
Is this ready for testing? I'm happy to dive in if so.

**Comment by pierotofy** [comment]
Not yet ready. Probably by end of week. Hopefully sooner.

File: opendm/config.py [hunk 2]
```diff
@@ -9,7 +9,7 @@
 import sys
 
 # parse arguments
-processopts = ['dataset', 'split', 'merge', 'opensfm', 'mve', 'odm_filterpoints',
+processopts = ['dataset', 'split', 'merge', 'opensfm', 'openmvs', 'odm_filterpoints',
```
Comment (smathermather) [inline opendm/config.py:12]: Squee! Ahem. I mean, an astute choice.

Comment (Saijin-Naib) [reply opendm/config.py:12]: OpenMVS does make some crazy-good meshes, huh?
https://arc-team-open-research.blogspot.com/2016/12/comparing-7-photogrammetry-systems.html

Comment (pierotofy) [reply opendm/config.py:12]: Currently it has the best F-score of all FOSS programs on the tanks and temples benchmark: https://www.tanksandtemples.org/leaderboard/?table_0-sort=my_avg_rank

**Comment by pierotofy** [comment]
Ready for testing! :+1:

**Comment by Saijin-Naib** [comment]
Waiting for a SNAP build to go off successfully.

How can I verify within WebODM that I'm on the right release to try the new hotness out?

**Comment by smathermather** [comment]
I am already appreciating OpenSfM's improved concurrency... . This is going to be fun.

**Comment by pierotofy** [comment]
@Saijin-Naib you need to build the docker image on your own to test:

```
git clone https://github.com/OpenDroneMap/ODM
cd ODM
git remote add piero https://github.com/pierotofy/ODM
git fetch piero 210
git checkout piero/210
docker build -t opendronemap/odm .
```

Then run from the command line (WebODM requires NodeODM, you'd need to rebuild that one too if you want to try with WebODM).

**Comment by pierotofy** [comment]
By the way, I still need to add better filtering, so if the point clouds look pretty OK now, wait to see after filtering...

anim

**Comment by smathermather** [comment]
Ah. That might explain some artifacts I was seeing in orthos, specifically along the edges of complicated buildings.

**Comment by pierotofy** [comment]
Now with filters!

**Comment by smathermather** [comment]
Rebuilt and running. In the meantime, if I add the depthmap flag rather than running with the default range, I get the following error (true for 1200 or 2400 as depthmap resolution):
```
2020-11-03 02:59:03,336 DEBUG: Undistorting image fort_270-45 (49).JPG
[INFO]    running /code/SuperBuild/src/opensfm/bin/opensfm export_visualsfm --points "/datasets/fort/opensfm"
[WARNING] Will skip exporting /datasets/fort/opensfm/geocoords_transformation.txt
[INFO]    Finished opensfm stage
[INFO]    Running openmvs stage
[INFO]    running /code/SuperBuild/src/opensfm/bin/opensfm export_openmvs "/datasets/fort/opensfm"
[INFO]    Running dense reconstruction. This might take a while.
[INFO]    running /code/SuperBuild/install/bin/OpenMVS/DensifyPointCloud "/datasets/fort/opensfm/undistorted/openmvs/scene.mvs"  --resolution-level 0 --min-resolution 2400.0 --max-resolution 3194 --max-threads 20 -w "/datasets/fort/opensfm/undistorted/openmvs/depthmaps" -v 0
03:03:23 [App     ] exception thrown: bad lexical cast: source type value could not be interpreted as target```

It could be

**Comment by pierotofy** [comment]
Mm, interesting. It works with the default though?

**Comment by pierotofy** [comment]
Well good news, I can reproduce it on a separate dataset :clinking_glasses: 

Ah, I bet it's a silly int < -- > float mismatch.

**Comment by pierotofy** [comment]
Fixed!

**Comment by pierotofy** [comment]
Before/after:

anim

Cars and buildings are especially improved.

**Comment by pierotofy** [comment]
Unless there are objections I will merge this by end of day.

**Comment by pierotofy** [comment]
Excellent results on the sand key dataset:

image

**Comment by Saijin-Naib** [comment]
Yowsa! These are so awesome! No joke, this is a game-changer.

Can you merge it now, pretty please, so I can stop obsessively refreshing this PR, and return to more healthy behavior like doomscrolling on twitter until I have a mental health crisis 🤣

**Comment by smathermather** [comment]
Has this been tested on split-merge, or should I run it through a test? Also, with an updated OpenSfM branch, do we have updated recommendations for large datasets? I seem to remember @YanNoun indicating there are some improvements in there for larger datasets.

**Comment by smathermather** [comment]
As with all noise reduction, most places improve, some detail is lost here and there:
noisereduction

**Comment by pierotofy** [comment]
I tested on some very small split-merge datasets, but nothing large. Always good to do more testing, especially with large changes like these.

**Comment by smathermather** [comment]
Cool. I'll run a couple k through first. It's a dataset I just tested on ODM 2.0.0, so it should make for a good direct comparison.

**Comment by smathermather** [comment]
Totally dumb question, I did an update on all my nodes (after confirming an update went through ~5 hours ago) but they are reporting still to be 2.0.0:
image

Did NodeODM not get a version bump?

**Comment by pierotofy** [comment]
NodeODM did not get a version bump. But you can check by visiting the nodes `http://ipofnode>:3000/info` and check that the "engineVersion" key is `2.1.0`.

**Comment by smathermather** [comment]
Perfect. Thanks!

**Comment by smathermather** [comment]
> As with all noise reduction, most places improve, some detail is lost here and there:

Of course, we can just bump up the depthmap resolution a smidge and all noise removal is forgiven:
image

**Comment by smathermather** [comment]
Possible bug in meshing for orthophoto. I suspect it has to do with the way we are calculating maximum extent and the the eagerness with which OpenSfM is calculating wider and better SfM solutions, but that's purely a hunch:
https://community.opendronemap.org/t/possible-bug-in-orthophoto-meshing-in-2-1-0/5853

**Comment by smathermather** [comment]
Ran a larger dataset through (about 8k images) split merge. It gets to the alignment step and fails silently:
```
[INFO]    LRE: Download of submodel_0003 at [99%]
[INFO]    LRE: Download of submodel_0003 at [99%]
[INFO]    LRE: Download of submodel_0003 at [99%]
[INFO]    LRE: Download of submodel_0003 at [100%]
[INFO]    LRE: Downloaded and extracted assets for submodel_0003
[INFO]    LRE: submodel_0003 finished successfully
[INFO]    LRE: Cleaning up remote task (6f52294d-52b2-4cd3-b375-47ef4613a9a3)... OK
[INFO]    LRE: No remote tasks left to cleanup
[INFO]    Aligning submodels...
```

If I rerun, it, it fails silently again:
```
gisuser@webodm:~$ docker logs optimistic_mahavira
[INFO]    DTM is turned on, automatically turning on point cloud classification
[INFO]    Initializing ODM - Fri Nov 06 14:27:40  2020
[INFO]    ==============
[INFO]    build_overviews: False
[INFO]    camera_lens: auto
[INFO]    cameras: {}
[INFO]    crop: 3
[INFO]    debug: False
[INFO]    dem_decimation: 1
[INFO]    dem_euclidean_map: False
[INFO]    dem_gapfill_steps: 3
[INFO]    dem_resolution: 5
[INFO]    depthmap_resolution: 1200.0
[INFO]    dsm: True
[INFO]    dtm: True
[INFO]    end_with: odm_report
[INFO]    fast_orthophoto: False
[INFO]    feature_quality: high
[INFO]    feature_type: sift
[INFO]    force_gps: False
[INFO]    gcp: None
[INFO]    geo: None
[INFO]    gps_accuracy: 10
[INFO]    ignore_gsd: False
[INFO]    matcher_distance: 0
[INFO]    matcher_neighbors: 8
[INFO]    max_concurrency: 20
[INFO]    merge: all
[INFO]    mesh_octree_depth: 10
[INFO]    mesh_point_weight: 4
[INFO]    mesh_samples: 1.0
[INFO]    mesh_size: 200000
[INFO]    min_num_features: 8000
[INFO]    name: znzcityfocus1
[INFO]    opensfm_depthmap_method: PATCH_MATCH
[INFO]    opensfm_depthmap_min_consistent_views: 3
[INFO]    opensfm_depthmap_min_patch_sd: 1
[INFO]    optimize_disk_space: False
[INFO]    orthophoto_compression: DEFLATE
[INFO]    orthophoto_cutline: False
[INFO]    orthophoto_no_tiled: False
[INFO]    orthophoto_png: False
[INFO]    orthophoto_resolution: 5
[INFO]    pc_classify: True
[INFO]    pc_csv: False
[INFO]    pc_ept: True
[INFO]    pc_filter: 2.5
[INFO]    pc_las: False
[INFO]    pc_rectify: False
[INFO]    pc_sample: 0
[INFO]    project_path: /datasets
[INFO]    radiometric_calibration: none
[INFO]    rerun: None
[INFO]    rerun_all: False
[INFO]    rerun_from: None
[INFO]    resize_to: 2048
[INFO]    skip_3dmodel: False
[INFO]    sm_cluster: True
[INFO]    smrf_scalar: 0.0
[INFO]    smrf_slope: 0.06
[INFO]    smrf_threshold: 1.05
[INFO]    smrf_window: 100.0
[INFO]    split: 1000
[INFO]    split_multitracks: False
[INFO]    split_overlap: 150
[INFO]    texturing_data_term: gmi
[INFO]    texturing_keep_unseen_faces: False
[INFO]    texturing_nadir_weight: 16
[INFO]    texturing_outlier_removal_type: gauss_clamping
[INFO]    texturing_skip_global_seam_leveling: False
[INFO]    texturing_skip_hole_filling: False
[INFO]    texturing_skip_local_seam_leveling: False
[INFO]    texturing_skip_visibility_test: False
[INFO]    texturing_tone_mapping: none
[INFO]    tiles: False
[INFO]    time: False
[INFO]    use_3dmesh: False
[INFO]    use_exif: False
[INFO]    use_fixed_camera_params: False
[INFO]    use_hybrid_bundle_adjustment: False
[INFO]    use_opensfm_dense: False
[INFO]    verbose: False
[INFO]    ==============
[INFO]    Running dataset stage
[INFO]    Loading dataset from: /datasets/znzcityfocus1/images
[INFO]    Loading images database: /datasets/znzcityfocus1/images.json
[INFO]    Found 7952 usable images
[INFO]    Coordinates file already exist: /datasets/znzcityfocus1/odm_georeferencing/coords.txt
[INFO]    Parsing SRS header: WGS84 UTM 37S
[INFO]    Finished dataset stage
[INFO]    Running split stage
[INFO]    Setting max-concurrency to 19 to better handle remote splits
[INFO]    Large dataset detected (7952 photos) and split set at 1000. Preparing split merge.
[WARNING] /datasets/znzcityfocus1/opensfm/image_list.txt already exists, not rerunning OpenSfM setup
[WARNING] Submodels directory already exist at: /datasets/znzcityfocus1/submodels
[INFO]    Dataset has been split into 8 submodels. Reconstructing each submodel...
[INFO]    LRE: Initializing using cluster node 10.0.0.17:3000
[INFO]    LRE: Node is online and running odm version 2.1.0
[INFO]    LRE: Adding to queue /datasets/znzcityfocus1/submodels/submodel_0000
[INFO]    LRE: Adding to queue /datasets/znzcityfocus1/submodels/submodel_0001
[INFO]    LRE: Adding to queue /datasets/znzcityfocus1/submodels/submodel_0002
[INFO]    LRE: Adding to queue /datasets/znzcityfocus1/submodels/submodel_0003
[INFO]    LRE: Adding to queue /datasets/znzcityfocus1/submodels/submodel_0004
[INFO]    LRE: Adding to queue /datasets/znzcityfocus1/submodels/submodel_0005
[INFO]    LRE: Adding to queue /datasets/znzcityfocus1/submodels/submodel_0006
[INFO]    LRE: Adding to queue /datasets/znzcityfocus1/submodels/submodel_0007
[INFO]    LRE: About to process submodel_0000 locally
[INFO]    ==================================
[INFO]    Local Reconstruction submodel_0000
[INFO]    ==================================
[INFO]    LRE: About to process submodel_0001 remotely
[WARNING] Detect features already done: /datasets/znzcityfocus1/submodels/submodel_0000/opensfm/features exists
[WARNING] Match features already done: /datasets/znzcityfocus1/submodels/submodel_0000/opensfm/matches exists
[WARNING] Found a valid OpenSfM tracks file in: /datasets/znzcityfocus1/submodels/submodel_0000/opensfm/tracks.csv
[WARNING] Found a valid OpenSfM reconstruction file in: /datasets/znzcityfocus1/submodels/submodel_0000/opensfm/reconstruction.json
[INFO]    Already processed feature matching and reconstruction for submodel_0001
[INFO]    LRE: submodel_0001 finished successfully
[INFO]    LRE: About to process submodel_0002 remotely
[INFO]    LRE: submodel_0000 finished successfully
[INFO]    LRE: About to process submodel_0003 locally
[INFO]    ==================================
[INFO]    Local Reconstruction submodel_0003
[INFO]    ==================================
[INFO]    Already processed feature matching and reconstruction for submodel_0002
[INFO]    LRE: submodel_0002 finished successfully
[INFO]    LRE: About to process submodel_0004 remotely
[WARNING] Detect features already done: /datasets/znzcityfocus1/submodels/submodel_0003/opensfm/features exists
[INFO]    Already processed feature matching and reconstruction for submodel_0004
[INFO]    LRE: submodel_0004 finished successfully
[INFO]    LRE: About to process submodel_0005 remotely
[WARNING] Match features already done: /datasets/znzcityfocus1/submodels/submodel_0003/opensfm/matches exists
[WARNING] Found a valid OpenSfM tracks file in: /datasets/znzcityfocus1/submodels/submodel_0003/opensfm/tracks.csv
[INFO]    Already processed feature matching and reconstruction for submodel_0005
[INFO]    LRE: submodel_0005 finished successfully
[INFO]    LRE: About to process submodel_0006 remotely
[WARNING] Found a valid OpenSfM reconstruction file in: /datasets/znzcityfocus1/submodels/submodel_0003/opensfm/reconstruction.json
[INFO]    Already processed feature matching and reconstruction for submodel_0006
[INFO]    LRE: submodel_0006 finished successfully
[INFO]    LRE: About to process submodel_0007 remotely
[INFO]    Already processed feature matching and reconstruction for submodel_0007
[INFO]    LRE: submodel_0007 finished successfully
[INFO]    LRE: submodel_0003 finished successfully
[INFO]    LRE: No remote tasks left to cleanup
[INFO]    Aligning submodels...
```

When I ran with a manually specified split-merge, I got the following error:
```
[INFO]    LRE: submodel_0004 finished successfully
[INFO]    LRE: No remote tasks left to cleanup
[INFO]    Aligning submodels...
Traceback (most recent call last):
  File "/code/run.py", line 69, in <module>
    app.execute()
  File "/code/stages/odm_app.py", line 86, in execute
    self.first_stage.run()
  File "/code/opendm/types.py", line 358, in run
    self.next_stage.run(outputs)
  File "/code/opendm/types.py", line 339, in run
    self.process(self.args, outputs)
  File "/code/stages/splitmerge.py", line 173, in process
    octx.align_reconstructions(self.rerun())
  File "/code/opendm/osfm.py", line 290, in align_reconstructions
    True)
  File "/code/SuperBuild/src/opensfm/opensfm/large/tools.py", line 244, in align_reconstructions
    reconstruction_name)
  File "/code/SuperBuild/src/opensfm/opensfm/large/tools.py", line 109, in add_camera_constraints_soft
    for shot_id in shots:
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfb in position 17: invalid start byte
```

**Comment by pierotofy** [comment]
Does it happen with small datasets too?

**Comment by smathermather** [comment]
Small datasets are like small talk:



In all seriousness, rerunning a 430 image datset with a split of 50. It shouldn't take too long.

edit: unless it fails. Running another one... .

**Comment by pierotofy** [comment]
Is the command to process the dataset run from Windows using a docker mounted volume by any chance?

**Comment by smathermather** [comment]
Noap. Docker on Ubuntu 18.04.5 LTS (GNU/Linux 4.15.0-122-generic x86_64). It is sitting on a VM on a VMWare instance.

**Comment by smathermather** [comment]
Split merge on a smaller dataset worked, but run through WebODM. Do you want me to run a smaller one on the filesystem from a docker instance for comparatives?

**Comment by pierotofy** [comment]
Yes, it's a strange error and the fact that smaller datasets are working OK (I just tested one myself also) is quite suspicious; the environment probably makes a difference. Reproducing the error consistently is also important (if it's a random/one-off error, I foresee a difficult troubleshooting session).

**Comment by smathermather** [comment]
Testing again with a few hundred images running from a docker ODM. Out of curiousity, did you run your small dataset on docker on the command line?

**Comment by pierotofy** [comment]
Yep, docker from the command line.

**Comment by smathermather** [comment]
Hmm. Well, I am transfering a smaller dataset for testing, but also I'll try a native run once I am upgraded to 20.04.

**Comment by smathermather** [comment]
Hmm, I upgraded to 20.04 and ~~it hasn't failed yet in rerunning~~ on the large dataset, ~~but~~ it does throw this warning at the start:
```
docker run -ti -v /home/gisuser/outdir/znz/images/:/datasets opendronemap/odm --project-path /datasets znzcityfocus1 --dtm --dsm --smrf-scalar 0 --smrf-slope 0.06 --split 1000 --smrf-threshold 1.05 --smrf-window 100 --depthmap-resolution 1200 --sm-cluster http://10.0.0.17:3000 --pc-ept
/code/SuperBuild/src/opensfm/opensfm/large/tools.py:242: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if camera_constraint_type is 'soft_camera_constraint':
/code/SuperBuild/src/opensfm/opensfm/large/tools.py:245: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if camera_constraint_type is 'hard_camera_constraint':
```

And then fails with this:
```
[INFO]    Aligning submodels...



Traceback (most recent call last):
  File "/code/run.py", line 69, in <module>
    app.execute()
  File "/code/stages/odm_app.py", line 86, in execute
    self.first_stage.run()
  File "/code/opendm/types.py", line 358, in run
    self.next_stage.run(outputs)
  File "/code/opendm/types.py", line 339, in run
    self.process(self.args, outputs)
  File "/code/stages/splitmerge.py", line 173, in process
    octx.align_reconstructions(self.rerun())
  File "/code/opendm/osfm.py", line 288, in align_reconstructions
    transformations = tools.align_reconstructions(reconstruction_shots,
  File "/code/SuperBuild/src/opensfm/opensfm/large/tools.py", line 243, in align_reconstructions
    add_camera_constraints_soft(ra, reconstruction_shots,
  File "/code/SuperBuild/src/opensfm/opensfm/large/tools.py", line 109, in add_camera_constraints_soft
    for shot_id in shots:
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x96 in position 1: invalid start byte
```

Addendum:
I haven't gotten the native install working yet, but I'll work on processing the smaller dataset.

**Comment by smathermather** [comment]
Instead I am testing with the full dataset in WebODM. I'll test with the smaller dataset on the command line if need be.

**Comment by smathermather** [comment]
It fails in WebODM too as follows:
```
[INFO]    LRE: Download of submodel_0004 at [99%]
[INFO]    LRE: Download of submodel_0004 at [100%]
[INFO]    LRE: Downloaded and extracted assets for submodel_0004
[INFO]    LRE: submodel_0004 finished successfully
[INFO]    LRE: Cleaning up remote task (928b219d-f797-471f-b6f9-6a31ce5ef82c)... OK
[INFO]    LRE: No remote tasks left to cleanup
[INFO]    Aligning submodels...
/code/run.sh: line 5:  1224 Segmentation fault      (core dumped) python3 $RUNPATH/run.py "$@"
```

**Comment by pierotofy** [comment]
I would put together a reproducible test case (dataset + description of environment + options used) and open a new issue so that we can try to reproduce this and hopefully have a fix.

**Comment by smathermather** [comment]
Will do.

**Pull request merged**